R/census_vectors.R

```diff
@@ -0,0 +1,221 @@
+#' list_census_vectors('CA16')
+list_census_vectors <- function(dataset, use_cache = FALSE, quiet = TRUE) {
+  cache_file <- cache_path(dataset, "_vectors.rda")
+  if (!use_cache || !file.exists(cache_file)) {
+    url <- paste0("https://censusmapper.ca/api/v1/vector_info/", dataset,
+                  ".csv")
+    response <- if (!quiet) {
+      message("Querying CensusMapper API for vectors data...")
+      httr::GET(url, httr::progress())
+    } else {
+      httr::GET(url)
+    }
+    handle_cm_status_code(response, NULL)
+    content <- httr::content(response, type = "text", encoding = "UTF-8")
+    result <- if (!requireNamespace("readr", quietly = TRUE)) {
+      dplyr::as_data_frame(utils::read.csv(textConnection(content), stringsAsFactors = FALSE))
+    } else {
+      readr::read_csv(content,col_types=readr::cols(.default="c"))
+    }
+    result <- dplyr::mutate(
+      result, type = factor(.data$type),
+      units = factor(units, levels = as.character(1:6),
+                     labels = c("Number", "Percentage ratio (0.0-1.0)",
+                                "Currency", "Ratio", "Percentage (0-100)","Currency (1000s)")),
+      aggregation = dplyr::case_when(
+        add == "1" ~ "Additive",
+        add == "0" ~ "Not additive",
+        grepl("^2.", add) ~ gsub(".", ", ", gsub("^2\\.", "Average of ", add),
+                                 fixed = TRUE),
+        grepl("^3.", add) ~ gsub(".", ", ", gsub("^3\\.", "Median of ", add),
+                                 fixed = TRUE),
+        grepl("^4.", add) ~ gsub(".", ", ", gsub("^4\\.", "Average to ", add),
+                                 fixed = TRUE)
+      )) %>%
+      dplyr::select(.data$vector, .data$type, .data$label, .data$units,
+                    parent_vector = .data$parent, .data$aggregation,
+                    .data$details)
+    attr(result, "last_updated") <- Sys.time()
+    attr(result, "dataset") <- dataset
+    save(result, file = cache_file)
+    result
+  } else {
+    if (!quiet) message("Reading vector information from local cache.")
+    load(file = cache_file)
+    last_updated <- attr(result, "last_updated")
+    if (!quiet && is.null(last_updated) ||
+        difftime(Sys.time(), last_updated, units = "days") > 1) {
+      warning(paste("Cached vectors list may be out of date. Set `use_cache =",
+                    "FALSE` to update it."))
+    }
+    attr(result, "dataset") <- dataset # just in case, catching cached legacy datasets
+    result
+  }
+}
+
+#' List all parent variables from vector hierarchical based on a (sub-)list of census
+#' variables returned by
+#' \code{list_census_vectors} or \code{search_census_vectors}.
+#'
+#' @param vector_list The list of vectors to be used
+#'
+#' @export
+#'
+#' @examples
+#' library(dplyr, warn.conflicts = FALSE)
+#'
+#' list_census_vectors("CA16") %>%
+#'   filter(vector == "v_CA16_4092") %>%
+#'   parent_census_vectors()
+parent_census_vectors <- function(vector_list){
```

Comment:
This should try to work the same way as child census vectors. Right now:

```
list_census_vectors("CA16") %>%
  filter(vector == "v_CA16_4092") %>%
  parent_census_vectors()
```
Works as usual, but
```
  parent_census_vectors("v_CA16_4092")
```
Throws a very odd error
```
Error in handle_cm_status_code(response, NULL) : 
  Download of Census Data failed. <!DOCTYPE html>
<html>
<head>
  <title>The page you were looking for doesn't exist (404)</title>
  <meta name="viewport" content="width=device-width,initial-scale=1">
  <style>
  body {
    background-color: #EFEFEF;
    color: #2E2F30;
    text-align: center;
    font-family: arial, sans-serif;
    margin: 0;
  }

  div.dialog {
    width: 95%;
    max-width: 33em;
    margin: 4em auto 0;
  }

  div.dialog > div {
    border: 1px solid #CCC;
    border-right-color: #999;
    border-left-color: #999;
    border-bottom-color: #BBB;
    border-top: #B00100 solid 4px;
    border-top-left-radius: 9px;
    border-top-right-radius: 9px;
    background-color: white;
    padding: 7px 12% 0;
    box-shadow: 0 3px 8px rgba(50, 50, 50, 0.17);
  }

  h1 {
    font-size: 100%;
    color: #730E15;
    line-height: 1.5em;
  }

  div.dialog > p {
    margin: 0 0 1em;
    padding: 1em;
    background-color: #F7F7F7;
    border: 1px solid #CCC;
    border-right-
```
If we don't want the same functionality, should at least add better error handling here.

Reply:
good catch, did not update the parent_census_vectors call. will do that.